tests/regression/test_mesh_vertexonly.py

```diff
@@ -0,0 +1,333 @@
+    for all MPI ranks to avoid hanging.
+    """
+    assert m.geometric_dimension() == gdim
+    # Correct dims
+    assert vm.geometric_dimension() == gdim
+    assert vm.topological_dimension() == 0
+    # Can initialise
+    vm.init()
+    # Find in-bounds and non-halo-region input coordinates
+    in_bounds = []
+    core, owned, ghost = m.cell_set.sizes
+    for i in range(len(inputvertexcoords)):
+        cell_num = m.locate_cell(inputvertexcoords[i])
+        if cell_num is not None and cell_num < owned:
+            in_bounds.append(i)
+    # Correct coordinates (though not guaranteed to be in same order)
+    assert np.shape(vm.coordinates.dat.data_ro) == np.shape(inputvertexcoords[in_bounds])
+    assert np.all(np.isin(inputvertexcoords[in_bounds], vm.coordinates.dat.data_ro))
+    # Correct parent topology
+    assert vm._parent_mesh is m
+    assert vm.topology._parent_mesh is m.topology
+    # Correct generic cell properties
+    assert np.shape(vm.cell_closure) == (len(inputvertexcoords[in_bounds]), 1)
+    with pytest.raises(AttributeError):
+        vm.cell_to_facets
+    assert vm.num_cells() == len(inputvertexcoords[in_bounds]) == vm.cell_set.size
+    assert vm.num_facets() == 0
+    assert vm.num_faces() == vm.num_entities(2) == 0
+    assert vm.num_edges() == vm.num_entities(1) == 0
+    assert vm.num_vertices() == vm.num_entities(0) == vm.num_cells()
+    # Correct parent cell numbers
+    stored_vertex_coords = np.copy(vm._swarm.getField("DMSwarmPIC_coor")).reshape((vm.num_cells(), gdim))
+    vm._swarm.restoreField("DMSwarmPIC_coor")
+    stored_parent_cell_nums = np.copy(vm._swarm.getField("parentcellnum"))
+    vm._swarm.restoreField("parentcellnum")
+    assert len(stored_vertex_coords) == len(stored_parent_cell_nums)
+    for i in range(len(stored_vertex_coords)):
+        assert m.locate_cell(stored_vertex_coords[i]) == stored_parent_cell_nums[i]
+
+
+@pytest.mark.parametrize("parentmesh", parentmeshes)
+def test_generate_cell_midpoints(parentmesh):
+    """
+    Generate cell midpoints for mesh m and check they lie in the correct cells
+    """
+    inputcoords, inputcoordslocal = cell_midpoints(parentmesh)
+    vm = VertexOnlyMesh(parentmesh, inputcoords)
+    # Midpoints located in correct cells of parent mesh
+    V = VectorFunctionSpace(parentmesh, "DG", 0)
+    f = Function(V).interpolate(parentmesh.coordinates)
+    # Check size of biggest len(vm.coordinates.dat.data_ro) so
+    # locate_cell can be called on every processor
+    max_len = MPI.COMM_WORLD.allreduce(len(vm.coordinates.dat.data_ro), op=MPI.SUM)
+    out_of_mesh_point = np.empty(shape=(1, parentmesh.geometric_dimension()))
+    out_of_mesh_point.fill(np.inf)
+    for i in range(max_len):
+        if i < len(vm.coordinates.dat.data_ro):
+            cell_num = parentmesh.locate_cell(vm.coordinates.dat.data_ro[i])
+        else:
+            cell_num = parentmesh.locate_cell(out_of_mesh_point)  # should return None
```

Comment:
We should promote the interface to locate_cell to provide `locate_cells` to reduce the amount of collective communicator required.

Actually, `locate_cell` is not collective other than the compilation step, which should probably be done on `COMM_SELF` instead.

Reply:
Switching to `COMM_SELF` (see diff below) sadly does not get rid of the need to do `locate_cell` on all processes.

```
diff --git a/firedrake/mesh.py b/firedrake/mesh.py
index 5c9d5bac..da8707c2 100644
--- a/firedrake/mesh.py
+++ b/firedrake/mesh.py
@@ -13,7 +13,7 @@ from mpi4py import MPI
 from pyop2.datatypes import IntType
 from pyop2 import op2
 from pyop2.base import DataSet
-from pyop2.mpi import COMM_WORLD, dup_comm
+from pyop2.mpi import COMM_WORLD, COMM_SELF, dup_comm
 from pyop2.profiling import timed_function, timed_region
 from pyop2.utils import as_tuple, tuplify
 
@@ -1349,7 +1349,8 @@ values from f.)"""
                                        + ["-I%s/include" % d for d in get_petsc_dir()],
                                        ldargs=["-L%s/lib" % sys.prefix,
                                                "-lspatialindex_c",
-                                               "-Wl,-rpath,%s/lib" % sys.prefix])
+                                               "-Wl,-rpath,%s/lib" % sys.prefix],
+                                        comm=COMM_SELF)
 
             locator.argtypes = [ctypes.POINTER(function._CFunction),
                                 ctypes.POINTER(ctypes.c_double)]
```

Reply:
Added #1756 in reference to this.